packages/network/src/network.js

```diff
@@ -27,554 +27,759 @@ export const rethrowUnlessMissing = err => {
-            ackDeferred.resolve(toBytes(ack));
-          },
-          err => {
-            pendingAcks.delete(ackDeferred);
-            ackDeferred.reject(err);
-          },
-        );
-      return ackDeferred.promise;
-    },
-  });
+export function getPrefixes(addr) {
+  const parts = addr.split(ENDPOINT_SEPARATOR);
 
-  current.add(connection);
-  E(handler)
-    .onOpen(connection, localAddr, remoteAddr, handler)
-    .catch(rethrowUnlessMissing);
-  return connection;
-};
+  /** @type {string[]} */
+  const ret = [];
+  for (let i = parts.length; i > 0; i -= 1) {
+    // Try most specific match.
+    const prefix = parts.slice(0, i).join(ENDPOINT_SEPARATOR);
+    ret.push(prefix);
+  }
+  return ret;
+}
 
 /**
- * @param {ConnectionHandler} handler0
- * @param {Endpoint} addr0
- * @param {ConnectionHandler} handler1
- * @param {Endpoint} addr1
- * @param {WeakSet<Connection>} [current]
- * @returns {[Connection, Connection]}
+ * @param {import('@agoric/base-zone').Zone} zone
+ * @param {ReturnType<import('@agoric/whenable').prepareWhenableModule>} powers
  */
-export function crossoverConnection(
-  handler0,
-  addr0,
-  handler1,
-  addr1,
-  current = new WeakSet(),
-) {
-  /** @type {Connection[]} */
-  const conns = [];
-  /** @type {ConnectionHandler[]} */
-  const handlers = [handler0, handler1];
-  /** @type {Endpoint[]} */
-  const addrs = [addr0, addr1];
+const prepareHalfConnection = (zone, { when }) => {
+  const makeHalfConnection = zone.exoClass(
+    'Connection',
+    undefined,
+    ({ addrs, handlers, conns, current, l, r }) => {
+      /** @type {Error | undefined} */
+      let closed;
 
-  function makeHalfConnection(l, r) {
-    let closed;
-    conns[l] = Far('Connection', {
+      return {
+        addrs,
+        handlers,
+        conns,
+        current,
+        l,
+        r,
+        closed,
+      };
+    },
+    {
       getLocalAddress() {
-        return addrs[l];
+        return this.state.addrs[this.state.l];
       },
       getRemoteAddress() {
-        return addrs[r];
+        return this.state.addrs[this.state.r];
       },
       async send(packetBytes) {
-        if (closed) {
-          throw closed;
+        if (this.state.closed) {
+          throw this.state.closed;
         }
-        const ack = await E(handlers[r])
-          .onReceive(conns[r], toBytes(packetBytes), handlers[r])
-          .catch(rethrowUnlessMissing);
+
+        // watch(
+        //   E(this.state.handlers[this.state.r])
+        //     .onReceive(
+        //       this.state.conns.get(this.state.r),
+        //       toBytes(packetBytes),
+        //       this.state.handlers[this.state.r],
+        //     )
+        //     .catch(rethrowUnlessMissing),
+        //   {
+        //     onFulfilled: ack => {
+
+        //     },
+        //     onRejected: () => { },
+        //   },
+        // );
+
+        const ack = await when(
+          E(this.state.handlers[this.state.r])
```

Comment:
This await _definitely_ needs to be robust in the presence of upgrade.
Hm. can we test that?

Also, the `await when(  E(this.state.protocolHandler)` around 575 in `prepareBinder`.

I hope to audit the code for others.

Reply:
Created [this](https://github.com/Agoric/agoric-sdk/issues/8867) to track that issue. This is prioritized as the next set of work after this PR ships.